component/loki/source/file/file.go

```diff
@@ -0,0 +1,277 @@
+	ForwardTo []chan api.Entry   `river:"forward_to,attr,optional"`
+}
+
+// DefaultArguments defines the default settings for loki.source.file.
+var DefaultArguments = Arguments{}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (arg *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*arg = DefaultArguments
+
+	type args Arguments
+	return f((*args)(arg))
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+)
+
+// Component implements the loki.source.file component.
+type Component struct {
+	opts    component.Options
+	metrics *metrics
+
+	mut       sync.RWMutex
+	args      Arguments
+	handler   chan api.Entry
+	receivers []chan api.Entry
+	posFile   positions.Positions
+	readers   map[string]Reader
+}
+
+// New creates a new loki.source.file component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	err := os.Mkdir(o.DataPath, 0750)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
+		SyncPeriod:        10 * time.Second,
+		PositionsFile:     filepath.Join(o.DataPath, "positions.yml"),
+		IgnoreInvalidYaml: false,
+		ReadOnly:          false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: newMetrics(o.Registerer),
+
+		handler:   make(chan api.Entry),
+		receivers: args.ForwardTo,
+		posFile:   positionsFile,
+		readers:   make(map[string]Reader),
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+// TODO(@tpaschalis). Should we periodically re-check? What happens if a target
+// comes alive _after_ it's been passed to us and we never receive another
+// Update()? Or should it be a responsibility of the discovery component?
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.file component shutting down, stopping readers")
+		for _, r := range c.readers {
+			r.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
+				receiver <- entry
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+	c.receivers = newArgs.ForwardTo
+
+	oldPaths := make(map[string]struct{})
+
+	// Stop all readers and recreate them below. This avoids the issue we saw
+	// with stranded wrapped handlers staying behind until they were GC'ed and
+	// sending duplicate message to the global handler. It also makes sure that
+	// we update everything with the new labels. Simply zeroing out the
+	// c.readers map did not work correctly to shut down the wrapped handlers
+	// in time.
```

Comment:
I'm assuming this comment is from Promtail, right? We might want to add a notice indicating as much. 

Recreating readers sounds expensive, so I'm interested in the future optimization, but the "wrapped handlers staying behind until they were GC'd and sending duplicate messages to the global handler" sounds a little scary. We'll probably want to learn more about this before doing the optimization.

Reply:
No, this is my own comment when I tried to explain the approach taken here.

After coming back to this today, I think it might be just as easy to compare the previous and new targets, and stop any readers that no longer exist or that had their labels changed, and re-start them accordingly, so I might include it in this PR.

Reply:
IMO if the below works reasonably well, I would go for checking this in and then revisiting in another PR. That way its more focused and gets this PR which imports Promtail out of the way.